pkg/provider/doppler/client/client.go

```diff
@@ -0,0 +1,348 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+type DopplerClient struct {
+	baseURL      *url.URL
+	DopplerToken string
+	VerifyTLS    bool
+	UserAgent    string
+}
+
+type queryParams map[string]string
+
+type headers map[string]string
+
+type httpRequestBody []byte
+
+type Secrets map[string]string
+
+type RawSecrets map[string]*interface{}
+
+type APIError struct {
+	Err     error
+	Message string
+	Data    string
+}
+
+type apiResponse struct {
+	HTTPResponse *http.Response
+	Body         []byte
+}
+
+type apiErrorResponse struct {
+	Messages []string
+	Success  bool
+}
+
+type SecretRequest struct {
+	Name    string
+	Project string
+	Config  string
+}
+
+type SecretsRequest struct {
+	Project         string
+	Config          string
+	NameTransformer string
+	Format          string
+	ETag            string // Specifying an ETag implies that the caller has implemented response caching
+}
+
+type UpdateSecretsRequest struct {
+	Secrets RawSecrets `json:"secrets,omitempty"`
+	Project string     `json:"project,omitempty"`
+	Config  string     `json:"config,omitempty"`
+}
+
+type secretResponseBody struct {
+	Name  string `json:"name,omitempty"`
+	Value struct {
+		Raw      *string `json:"raw"`
+		Computed *string `json:"computed"`
+	} `json:"value,omitempty"`
+	Messages *[]string `json:"messages,omitempty"`
+	Success  bool      `json:"success"`
+}
+
+type SecretResponse struct {
+	Name  string
+	Value string
+}
+
+type SecretsResponse struct {
+	Secrets  Secrets
+	Body     []byte
+	Modified bool
+	ETag     string
+}
+
+func NewDopplerClient(dopplerToken string) (*DopplerClient, error) {
+	client := &DopplerClient{
+		DopplerToken: dopplerToken,
+		VerifyTLS:    true,
+		UserAgent:    "doppler-external-secrets",
+	}
+
+	if err := client.SetBaseURL("https://api.doppler.com"); err != nil {
+		return nil, &APIError{Err: err, Message: "setting base URL failed"}
+	}
+
+	return client, nil
+}
+
+func (c *DopplerClient) BaseURL() *url.URL {
+	u := *c.baseURL
+	return &u
+}
+
+func (c *DopplerClient) SetBaseURL(urlStr string) error {
+	baseURL, err := url.Parse(strings.TrimSuffix(urlStr, "/"))
+
+	if err != nil {
+		return err
+	}
+
+	if baseURL.Scheme == "" {
+		baseURL.Scheme = "https"
+	}
+
+	c.baseURL = baseURL
+	return nil
+}
+
+func (c *DopplerClient) Authenticate() error {
```

Comment:
Need this in order to validate the Doppler provider. Projects endpoint used as it works without configuring the request for service and personal tokens.

Reply:
Will be a good thing to update when we add the `/me` endpoint.